samcli/lib/build/app_builder.py

```diff
@@ -191,8 +191,18 @@ def _get_build_graph(
 
         for function in functions:
             container_env_vars = self._make_env_vars(function, file_env_vars, inline_env_vars)
+
+            # In case of CI env, the project might have inconsistent absolute path,
+            # which will cause build cache to be invalidated.
+            # Here relative paths are always used in FunctionBuildDefinition (#2872).
+            codeuri = (
+                os.path.relpath(function.codeuri)
+                # use os.path.isabs() to opt-out URLs like file:///...
+                if isinstance(function.codeuri, str) and os.path.isabs(function.codeuri)
+                else function.codeuri
+            )
             function_build_details = FunctionBuildDefinition(
-                function.runtime, function.codeuri, function.packagetype, function.metadata, env_vars=container_env_vars
+                function.runtime, codeuri, function.packagetype, function.metadata, env_vars=container_env_vars
```

Comment:
Not totally sure, but having absolute path vs relative path might affect the build process. Can we store another field for relative path and use that one for validating the cache?

Reply:
I forgot to bring that up. Before nested stack feature was introduced, the codeuri here was relative paths. (verified with 1.19.0)

In nested stack feature, codeuri needs to be resolved relative against its template file, and `click` always convert template file path to absolute paths, which led to the current behavior